File: app/src/main/java/com/candy/capture/adapter/ContentListAdapter.java
package com.candy.capture.adapter;

import android.content.Context;
import android.graphics.Bitmap;
import android.media.ThumbnailUtils;
import android.provider.MediaStore;
import android.support.v4.app.Fragment;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.util.TimeUtils;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;

import com.candy.capture.R;
import com.candy.capture.activity.MainActivity;
import com.candy.capture.core.ConstantValues;
import com.candy.capture.customview.AudioView;
import com.candy.capture.model.Content;
import com.candy.capture.model.MediaPlayState;
import com.candy.capture.util.GlideImageLoader;
import com.candy.capture.util.TimeUtil;

import java.util.ArrayList;

/**
 * Created by zhanghq on 2016/10/30.
 */

public class ContentListAdapter extends RecyclerView.Adapter<ContentListAdapter.BaseViewHolder> {
    private Context mContext;
    private ArrayList<Content> mContents;
    private ItemClickCallback mItemClickCallback;
    private AudioPlayCallBack mAudioPlayCallBack;
    private boolean mIsEditMode;

    public ContentListAdapter(Context context, ArrayList<Content> contents, AudioPlayCallBack audioPlayCallBack, ItemClickCallback callback) {
        mContext = context;
        mContents = contents;
        mAudioPlayCallBack = audioPlayCallBack;
        mItemClickCallback = callback;
    }

//    public void setContents(ArrayList<Content> contents) {
//        mContents = contents;
//        notifyDataSetChanged();
//    }

//    public void appendContentsEnd(ArrayList<Content> contents) {
//        if (mContents == null) {
//            mContents = contents;
//            notifyDataSetChanged();
//        } else {
//            int startPosition = mContents.size() - 1;
//            int count = contents.size();
//            mContents.addAll(contents);
//            notifyItemRangeInserted(startPosition, count);
//        }
//    }
//
//    public void appendContentsStart(ArrayList<Content> contents) {
//        if (mContents == null) {
//            mContents = contents;
//            notifyDataSetChanged();
//        } else {
//            int count = contents.size();
//            mContents.addAll(0, contents);
//            notifyItemRangeInserted(0, count);
//        }
//    }

    public void setEditMode(boolean isEditMode) {
        mIsEditMode = isEditMode;
    }

    @Override
    public int getItemViewType(int position) {
        return mContents.get(position).getType();
    }

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        BaseViewHolder holder = null;
        switch (viewType) {
            case ConstantValues.CONTENT_TYPE_TEXT:
                holder = new TextViewHolder(View.inflate(mContext, R.layout.item_text, null));
                break;
            case ConstantValues.CONTENT_TYPE_AUDIO:
                holder = new AudioViewHolder(View.inflate(mContext, R.layout.item_audio, null));
                break;
            case ConstantValues.CONTENT_TYPE_PHOTO:
                holder = new ImageViewHolder(View.inflate(mContext, R.layout.item_image, null));
                break;
            case ConstantValues.CONTENT_TYPE_VIDEO:
                holder = new VideoViewHolder(View.inflate(mContext, R.layout.item_video, null));
                break;
        }
        return holder;
    }

    @Override
    public void onBindViewHolder(BaseViewHolder holder, final int position) {
        Content content = mContents.get(position);
        holder.mRootView.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                if (mItemClickCallback != null) {
                    mItemClickCallback.onItemLongPressed(position);
                }
                return true;
            }
        });
        holder.mRootView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mItemClickCallback != null) {
                    mItemClickCallback.onItemClick(position);
                }
            }
        });
        holder.mCheckBox.setVisibility(mIsEditMode ? View.VISIBLE : View.GONE);
        holder.mCheckBox.setChecked(content.isSelect());
        holder.mDescTv.setText(content.getDesc());

        if (holder instanceof AudioViewHolder) {
            final AudioViewHolder viewHolder = (AudioViewHolder) holder;
            viewHolder.mDurationTv.setText(TimeUtil.formatDuration(content.getMediaDuration()));
            viewHolder.mAudioView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mItemClickCallback != null) {
                        mItemClickCallback.onAudioViewClick(position);
                    }
                }
            });
            viewHolder.mAudioView.setDuration(content.getMediaDuration());
            if (content.isPlayingAudio()) {
                if (mAudioPlayCallBack != null) {
                    double percent = mAudioPlayCallBack.getPlayPercent();
                    if (percent < 0) {
                        viewHolder.mAudioView.stopAnimation();
                    } else {
                        viewHolder.mAudioView.startAnimation(percent);
                    }
                }
            } else {
                viewHolder.mAudioView.stopAnimation();
            }
        } else if (holder instanceof ImageViewHolder) {
            ImageViewHolder viewHolder = (ImageViewHolder) holder;
            viewHolder.mImageView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mItemClickCallback != null) {
                        mItemClickCallback.onImageViewClick(position);
                    }
                }
            });
            GlideImageLoader.getInstance().loadImage(mContext, content.getMediaFilePath(), viewHolder.mImageView);
        } else if (holder instanceof VideoViewHolder) {
            VideoViewHolder viewHolder = (VideoViewHolder) holder;
            GlideImageLoader.getInstance().loadVideoThumbnail(mContext, content.getMediaFilePath(), viewHolder.mImageView);
            viewHolder.mPlaybackMask.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mItemClickCallback != null) {
                        mItemClickCallback.onVideoViewClick(position);
                    }
                }
            });
        }
    }

    @Override
    public int getItemCount() {
        return mContents == null ? 0 : mContents.size();
    }

    class BaseViewHolder extends RecyclerView.ViewHolder {
        protected View mRootView;
        protected TextView mDescTv;
        protected CheckBox mCheckBox;

        public BaseViewHolder(View itemView) {
            super(itemView);

            mRootView = itemView.findViewById(R.id.rl_root);
            mDescTv = (TextView) itemView.findViewById(R.id.tv_desc);
            mCheckBox = (CheckBox) itemView.findViewById(R.id.cb_select);
        }
    }

    class TextViewHolder extends BaseViewHolder {

        public TextViewHolder(View itemView) {
            super(itemView);
        }
    }

    class AudioViewHolder extends BaseViewHolder {
        private AudioView mAudioView;
        private TextView mDurationTv;

        public AudioViewHolder(View itemView) {
            super(itemView);

            mAudioView = (AudioView) itemView.findViewById(R.id.audio_view);
            mDurationTv = (TextView) itemView.findViewById(R.id.tv_duration);
        }
    }

    class ImageViewHolder extends BaseViewHolder {
        private ImageView mImageView;

        public ImageViewHolder(View itemView) {
            super(itemView);

            mImageView = (ImageView) itemView.findViewById(R.id.image_view);
        }
    }

    class VideoViewHolder extends BaseViewHolder {
        private ImageView mImageView;
        private ImageView mPlaybackMask;

        public VideoViewHolder(View itemView) {
            super(itemView);

            mImageView = (ImageView) itemView.findViewById(R.id.image_view);
            mPlaybackMask = (ImageView) itemView.findViewById(R.id.iv_playback_mask);
        }
    }

    public interface ItemClickCallback {
        void onItemClick(int position);

        void onItemLongPressed(int position);

        void onAudioViewClick(int position);

        void onImageViewClick(int position);

        void onVideoViewClick(int position);
    }

    public interface AudioPlayCallBack {
        double getPlayPercent();
    }
}

File: app/src/main/java/com/candy/capture/adapter/ContentListAdapter.kt
package com.candy.capture.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.databinding.ViewDataBinding
import androidx.databinding.library.baseAdapters.BR
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.candy.capture.R
import com.candy.capture.core.ConstantValues
import com.candy.capture.customview.AudioView
import com.candy.capture.databinding.ItemAudioBinding
import com.candy.capture.databinding.ItemImageBinding
import com.candy.capture.databinding.ItemTextBinding
import com.candy.capture.databinding.ItemVideoBinding
import com.candy.capture.model.Content
import com.candy.commonlibrary.utils.GlideImageLoader
import java.util.*


/**
 * @Description
 * @Author zhanghaiqiang
 * @Date 2020/9/30 11:13
 */
class ContentListAdapter(val context: Context?,
                         private val audioPlayCallBack: AudioPlayCallBack?) : PagedListAdapter<Content, ContentListAdapter.BaseViewHolder>(DIFF_CALLBACK) {
    companion object {
        private val DIFF_CALLBACK = object :
                DiffUtil.ItemCallback<Content>() {
            // Content details may have changed if reloaded from the database,
            // but ID is fixed.
            override fun areItemsTheSame(oldConcert: Content,
                                         newConcert: Content) = oldConcert.id == newConcert.id

            override fun areContentsTheSame(oldConcert: Content,
                                            newConcert: Content) = oldConcert == newConcert
        }
    }

    private var mIsEditMode = false
    var callbackContent: ItemClickContentCallback? = null

    fun setEditMode(isEditMode: Boolean) {
        mIsEditMode = isEditMode
    }

    override fun getItemViewType(position: Int): Int {
        return currentList!![position]?.type!!
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        return when (viewType) {
            ConstantValues.CONTENT_TYPE_TEXT -> TextViewHolder(ItemTextBinding.inflate(layoutInflater))
            ConstantValues.CONTENT_TYPE_AUDIO -> AudioViewHolder(ItemAudioBinding.inflate(layoutInflater))
            ConstantValues.CONTENT_TYPE_PHOTO -> ImageViewHolder(ItemImageBinding.inflate(layoutInflater))
            ConstantValues.CONTENT_TYPE_VIDEO -> VideoViewHolder(ItemVideoBinding.inflate(layoutInflater))
            else -> TextViewHolder(ItemTextBinding.inflate(layoutInflater))
        }
    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
        val content = currentList!![position]!!
        if (holder is AudioViewHolder) {
//            holder.mAudioView.setDuration(content.mediaDuration)
//            if (content.isPlayingAudio) {
//                if (audioPlayCallBack != null) {
//                    val percent = audioPlayCallBack.getPlayPercent()
//                    if (percent < 0) {
//                        holder.mAudioView.stopAnimation()
//                    } else {
//                        holder.mAudioView.startAnimation(percent)
//                    }
//                }
//            } else {
//                holder.mAudioView.stopAnimation()
//            }
        } else if (holder is ImageViewHolder) {
//            GlideImageLoader.getInstance().loadImage(context!!, content.mediaFilePath!!, holder.mImageView)
        } else if (holder is VideoViewHolder) {
            GlideImageLoader.getInstance().loadVideoThumbnail(context!!, content.mediaFilePath, holder.mImageView)
        }
        holder.bindTo(content)
    }

    open inner class BaseViewHolder(private val binding: ViewDataBinding) : RecyclerView.ViewHolder(binding.root) {
//        var mDescTv: TextView = itemView.findViewById<View>(R.id.tv_desc) as TextView
//        var mCheckBox: CheckBox = itemView.findViewById<View>(R.id.cb_select) as CheckBox

        open fun bindTo(content: Content) {
            binding.setVariable(BR.content, content)
            binding.setVariable(BR.clickHandler, callbackContent)
            binding.setVariable(BR.editMode, mIsEditMode)
            binding.setVariable(BR.audioPlayer, audioPlayCallBack)
            binding.executePendingBindings()
        }
    }

    internal inner class TextViewHolder(private val binding: ViewDataBinding) : BaseViewHolder(binding) {
//        override fun bindTo(content: Content) {
//            super.bindTo(content)
//            binding.executePendingBindings()
//        }
    }

    internal inner class AudioViewHolder(private val binding: ViewDataBinding) : BaseViewHolder(binding) {
        val mAudioView: AudioView = itemView.findViewById<View>(R.id.audio_view) as AudioView
//        val mDurationTv: TextView = itemView.findViewById<View>(R.id.tv_duration) as TextView

//        override fun bindTo(content: Content) {
//            super.bindTo(content)
//            binding.setVariable()
//            binding.executePendingBindings()
//        }
    }

    internal inner class ImageViewHolder(private val binding: ViewDataBinding) : BaseViewHolder(binding) {
//        val mImageView: ImageView = itemView.findViewById<View>(R.id.image_view) as ImageView

    }

    internal inner class VideoViewHolder(private val binding: ViewDataBinding) : BaseViewHolder(binding) {
        val mImageView: ImageView = itemView.findViewById<View>(R.id.image_view) as ImageView
//        val mPlaybackMask: ImageView = itemView.findViewById<View>(R.id.iv_playback_mask) as ImageView

    }

    interface ItemClickContentCallback {
        fun onItemClick(content: Content)
        fun onItemLongPressed(content: Content): Boolean
        fun onAudioViewClick(content: Content)
        fun onImageViewClick(content: Content)
        fun onVideoViewClick(content: Content)
    }

    interface AudioPlayCallBack {
        fun getPlayPercent(): Double
    }
}